pkg/http/tripperware/tripperware.go

```diff
@@ -34,3 +35,10 @@ func DefaultWithTransport(transport *http.Transport, config *tls.Config) http.Ro
 	transport.TLSClientConfig = config
 	return &defaultTripper{Transport: transport}
 }
+
+func Close(r io.Closer) error {
```

Comment:
An alternative would be to have a slightly more complex helper like this:
```
func toNext(req *http.Request, rt http.RoundTripper) (http.RoundTripper, func()) {
	if req.Body == nil {
		return rt, func() {}
	}
	forwarded := false
	return httpwares.RoundTripperFunc(func(actualReq *http.Request) (*http.Response, error) {
		if actualReq.Body == req.Body {
			forwarded = true
		}
		return rt.RoundTrip(actualReq)
	}), func() {
		if !forwarded {
			_ = req.Body.Close()
		}
	}
}
```

used like this in a tripperware:
```
func (t tripper) RoundTrip(req *http.Request) (*http.Response, error) {
  next, close := toNext(req)
  defer close()
  ...
  return next(req)
}
```

Reply:
Agree, also I don't like this function. For something that small it should be just copied also named `closedIfNotNil` or something

Reply:
your helper makes sense just maybe `ensureBodyClosed` would be more precise name?

Reply:
Yep naming was a bit quick'n'dirty. `EnsureBodyClosed` sounds ok.

Reply:
So we have multiple tripperwares. Like 6-7. With this we essentially double the stack frame as single tripperware is now actually 2 - the tripperware + EnsureBodyClosed. Even though we don't care about number of tripperwares this feels wrong. Maybe some wrapper over body in single place to close it if not closed on exit path?

Reply:
Wrapping body cleanly is hard, as the concrete type may implement various interfaces for optimisation (readerfrom, writerto). I don't get the same sense of this feeling wrong, can you elaborate why?

Reply:
I decided to just go for simple explicit closes, what do you think? I just think it's too much code for something that simple ):

Reply:
To me, the explicit closes feel significantly more error-prone than the other helper, so more likely to be gotten wrong in future changes. Consider that we got this wrong at first, and that future developers will need to read through the http request docs the same way we have to understand the rationale for the close calls. Maybe there's another way, maybe `closeBody(*http.Request)` will make it clearer?